

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Run your own model &mdash; DrEvalPy 1.3.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/custom_cookietemple.css?v=a7e27724" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=bb516dca"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Contributor Guide" href="contributing.html" />
    <link rel="prev" title="How to use DrEvalPy" href="usage.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html">
            
              <img src="_static/drevalpy.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">How to use DrEvalPy</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Run your own model</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#example-proteomics-random-forest">Example: Proteomics Random Forest</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributor Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference.html">How to cite</a></li>
<li class="toctree-l1"><a class="reference internal" href="API.html">API</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">DrEvalPy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Run your own model</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/daisybio/drevalpy/blob/development/docs/runyourmodel.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="run-your-own-model">
<h1>Run your own model<a class="headerlink" href="#run-your-own-model" title="Link to this heading"></a></h1>
<p>DrEvalPy provides a standardized interface for running your own model.
First make a new folder for your model at <code class="docutils literal notranslate"><span class="pre">drevalpy/models/your_model_name</span></code>.
Create <code class="docutils literal notranslate"><span class="pre">drevalpy/models/your_model_name/your_model.py</span></code> in which you need to define the Python class for your model.
This class should inherit from the <a class="reference internal" href="drevalpy.models.html#drp-label"><span class="std std-ref">DRPModel</span></a> base class.
DrEvalPy is agnostic to the specific modeling strategy you use. However, you need to define the input views, which represent different types of features that your model requires.
In this example, the model uses “gene_expression” and “methylation” features as cell line views, and “fingerprints” as drug views.
Additionally, you must define a unique model name to identify your model during evaluation.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">drevalpy.models.drp_model</span><span class="w"> </span><span class="kn">import</span> <span class="n">DRPModel</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">drevalpy.datasets.dataset</span><span class="w"> </span><span class="kn">import</span> <span class="n">FeatureDataset</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="k">class</span><span class="w"> </span><span class="nc">YourModel</span><span class="p">(</span><span class="n">DRPModel</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;A revolutionary new modeling strategy.&quot;&quot;&quot;</span>

    <span class="n">is_single_drug_model</span> <span class="o">=</span> <span class="kc">True</span> <span class="o">/</span> <span class="kc">False</span> <span class="c1"># TODO: set to true if your model is a single drug model</span>
    <span class="n">early_stopping</span> <span class="o">=</span> <span class="kc">True</span> <span class="o">/</span> <span class="kc">False</span> <span class="c1"># TODO: set to true if you want to use a part of the validation set for early stopping</span>
    <span class="n">cell_line_views</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gene_expression&quot;</span><span class="p">,</span> <span class="s2">&quot;methylation&quot;</span><span class="p">]</span>
    <span class="n">drug_views</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;fingerprints&quot;</span><span class="p">]</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">get_model_name</span><span class="p">(</span><span class="bp">cls</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the name of the model.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s2">&quot;YourModel&quot;</span>
</pre></div>
</div>
<p>Next let’s implement the feature loading. You have to return a DrEvalPy FeatureDataset object which contains the features for the cell lines and drugs.
If the features are different depending on the dataset, use the <code class="docutils literal notranslate"><span class="pre">dataset_name</span></code> parameter.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">load_drug_features</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">dataset_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">FeatureDataset</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Loads the drug features, in this case the drug ids.</span>

<span class="sd">    :param data_path: path to the data</span>
<span class="sd">    :param dataset_name: name of the dataset</span>
<span class="sd">    :returns: FeatureDataset containing the drug ids</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">feature_dataset</span> <span class="o">=</span> <span class="n">FeatureDataset</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data_path</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s2">/fingerprints.csv&quot;</span><span class="p">)</span> <span class="c1"># make sure to adjust the path to your data</span>

    <span class="k">return</span> <span class="n">feature_dataset</span>


<span class="k">def</span><span class="w"> </span><span class="nf">load_cell_line_features</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">dataset_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">FeatureDataset</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Loads the cell line features, in this case the cell line ids.</span>

<span class="sd">    :param data_path: path to the data</span>
<span class="sd">    :param dataset_name: name of the dataset</span>
<span class="sd">    :returns: FeatureDataset containing the cell line ids</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">feature_dataset</span> <span class="o">=</span> <span class="n">FeatureDataset</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data_path</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s2">_gene_expression.csv&quot;</span><span class="p">,</span>
                                                <span class="n">id_column</span><span class="o">=</span><span class="s2">&quot;cell_line_ids&quot;</span><span class="p">,</span>
                                                <span class="n">view_name</span><span class="o">=</span><span class="s2">&quot;gene_expression&quot;</span>
                                             <span class="p">)</span> <span class="c1"># make sure to adjust the path to your data</span>
    <span class="n">methylation</span> <span class="o">=</span> <span class="n">FeatureDataset</span><span class="o">.</span><span class="n">from_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data_path</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s2">_methylation.csv&quot;</span><span class="p">,</span>
                                            <span class="n">id_column</span><span class="o">=</span><span class="s2">&quot;cell_line_ids&quot;</span><span class="p">,</span>
                                            <span class="n">view_name</span><span class="o">=</span><span class="s2">&quot;gene_expression&quot;</span>
                                         <span class="p">)</span> <span class="c1"># make sure to adjust the path to your data</span>
    <span class="n">feature_dataset</span><span class="o">.</span><span class="n">add_features</span><span class="p">(</span><span class="n">methylation</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">feature_dataset</span>
</pre></div>
</div>
<p>The build_model functions can be used if you want to use tunable hyperparameters.
The hyperparameters which get tested are defined in the <code class="docutils literal notranslate"><span class="pre">drevalpy/models/your_model_name/hyperparameters.yaml</span></code>.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">build_model</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">hyperparameters</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Builds the model, for models that use hyperparameters.</span>

<span class="sd">    :param hyperparameters: hyperparameters for the model</span>
<span class="sd">    Example:</span>
<span class="sd">        self.model = ElasticNet(alpha=hyperparameters[&quot;alpha&quot;], l1_ratio=hyperparameters[&quot;l1_ratio&quot;])</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">predictor</span> <span class="o">=</span> <span class="n">YourPredictor</span><span class="p">(</span><span class="n">hyperparameters</span><span class="p">)</span> <span class="c1"># Initialize your Predictor, this could be a sklearn model, a neural network, etc.</span>
</pre></div>
</div>
<p>Sometimes, the model design is dependent on your training data input. In this case, you can also consider implementing build_model like:</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">build_model</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">hyperparameters</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">hyperparameters</span> <span class="o">=</span> <span class="n">hyperparameters</span>
</pre></div>
</div>
<p>and then set the model design later in the train method when you have access to the training data.
The train method should handle model training, and saving any necessary information (e.g., learned parameters).
Here we use a simple predictor that just uses the concatenated features to predict the response.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">train</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output</span><span class="p">:</span> <span class="n">DrugResponseDataset</span><span class="p">,</span> <span class="n">cell_line_input</span><span class="p">:</span> <span class="n">FeatureDataset</span><span class="p">,</span> <span class="n">drug_input</span><span class="p">:</span> <span class="n">FeatureDataset</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">output_earlystopping</span><span class="p">:</span> <span class="n">DrugResponseDataset</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>

    <span class="n">inputs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_feature_matrices</span><span class="p">(</span>
        <span class="n">cell_line_ids</span><span class="o">=</span><span class="n">cell_line_ids</span><span class="p">,</span>
        <span class="n">drug_ids</span><span class="o">=</span><span class="n">drug_ids</span><span class="p">,</span>
        <span class="n">cell_line_input</span><span class="o">=</span><span class="n">cell_line_input</span><span class="p">,</span>
        <span class="n">drug_input</span><span class="o">=</span><span class="n">drug_input</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">predictor</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="o">**</span><span class="n">inputs</span><span class="p">,</span> <span class="n">output</span><span class="o">.</span><span class="n">response</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">predictor</span> <span class="o">=</span> <span class="n">predictor</span> <span class="c1"># save your predictor for the prediction step</span>
</pre></div>
</div>
<p>In case you want to set some parameters dependent on the training data, your train function might look like this:</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">train</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output</span><span class="p">:</span> <span class="n">DrugResponseDataset</span><span class="p">,</span> <span class="n">cell_line_input</span><span class="p">:</span> <span class="n">FeatureDataset</span><span class="p">,</span> <span class="n">drug_input</span><span class="p">:</span> <span class="n">FeatureDataset</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">output_earlystopping</span><span class="p">:</span> <span class="n">DrugResponseDataset</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>

    <span class="n">cell_line_input</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_feature_selection</span><span class="p">(</span><span class="n">output</span><span class="p">,</span> <span class="n">cell_line_input</span><span class="p">)</span>
    <span class="n">dim_gex</span><span class="p">,</span> <span class="n">dim_mut</span><span class="p">,</span> <span class="n">dim_cnv</span> <span class="o">=</span> <span class="n">get_dimensions_of_omics_data</span><span class="p">(</span><span class="n">cell_line_input</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">nn_model</span> <span class="o">=</span> <span class="n">YourModel</span><span class="p">(</span>
                            <span class="n">input_size_gex</span><span class="o">=</span><span class="n">dim_gex</span><span class="p">,</span>
                            <span class="n">input_size_mut</span><span class="o">=</span><span class="n">dim_mut</span><span class="p">,</span>
                            <span class="n">input_size_cnv</span><span class="o">=</span><span class="n">dim_cnv</span><span class="p">,</span>
                            <span class="n">hpams</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">hyperparameters</span><span class="p">,</span>
                            <span class="o">...</span>
                        <span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">nn_model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span>
        <span class="n">output_train</span><span class="o">=</span><span class="n">output</span><span class="p">,</span>
        <span class="n">output_early_stopping</span><span class="o">=</span><span class="n">output_earlystopping</span><span class="p">,</span>
        <span class="n">cell_line_input</span><span class="o">=</span><span class="n">cell_line_input</span><span class="p">,</span>
        <span class="n">drug_input</span><span class="o">=</span><span class="n">drug_input</span><span class="p">,</span>
    <span class="p">)</span>
</pre></div>
</div>
<p>The predict method should handle model prediction, and return the predicted response values.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">predict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cell_line_input</span><span class="p">:</span> <span class="n">FeatureDataset</span><span class="p">,</span> <span class="n">drug_input</span><span class="p">:</span> <span class="n">FeatureDataset</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">:</span>

    <span class="n">inputs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_feature_matrices</span><span class="p">(</span>
        <span class="n">cell_line_ids</span><span class="o">=</span><span class="n">cell_line_ids</span><span class="p">,</span>
        <span class="n">drug_ids</span><span class="o">=</span><span class="n">drug_ids</span><span class="p">,</span>
        <span class="n">cell_line_input</span><span class="o">=</span><span class="n">cell_line_input</span><span class="p">,</span>
        <span class="n">drug_input</span><span class="o">=</span><span class="n">drug_input</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">predictor</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="o">**</span><span class="n">inputs</span><span class="p">,</span> <span class="n">output</span><span class="o">.</span><span class="n">response</span><span class="p">)</span>
</pre></div>
</div>
<p>Finally, you need to register your model with the framework. This can be done by adding the following line to the <code class="docutils literal notranslate"><span class="pre">__init__.py</span></code> file in the <code class="docutils literal notranslate"><span class="pre">drevalpy/models/__init__.py</span></code> directory.
Update the <code class="docutils literal notranslate"><span class="pre">MULTI_DRUG_MODEL_FACTORY</span></code> if your model is a global model for multiple cancer drugs or to the <code class="docutils literal notranslate"><span class="pre">SINGLE_DRUG_MODEL_FACTORY</span></code> if your model is specific to a single drug and needs to be trained for each drug separately.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">.your_model_name.your_model</span><span class="w"> </span><span class="kn">import</span> <span class="n">YourModel</span>
<span class="n">MULTI_DRUG_MODEL_FACTORY</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="s2">&quot;YourModel&quot;</span><span class="p">:</span> <span class="n">YourModel</span><span class="p">)</span>
</pre></div>
</div>
<p>Next, please also write appropriate tests in <code class="docutils literal notranslate"><span class="pre">tests/individual_models</span></code> and documentation in <code class="docutils literal notranslate"><span class="pre">docs/</span></code>.</p>
<section id="example-proteomics-random-forest">
<h2>Example: Proteomics Random Forest<a class="headerlink" href="#example-proteomics-random-forest" title="Link to this heading"></a></h2>
<p>Instead of gene expression data, we want to use proteomics data in our Random Forest.
The Random Forest is already implemented in <code class="docutils literal notranslate"><span class="pre">models/baselines/sklearn_models.py</span></code>.
We just need to adapt some methods.</p>
<p>1. We make a new class ProteomicsRandomForest which inherits from the RandomForest class.
We overwrite <code class="docutils literal notranslate"><span class="pre">cell_line_views</span></code> to <code class="docutils literal notranslate"><span class="pre">[&quot;proteomics&quot;]</span></code> and <code class="docutils literal notranslate"><span class="pre">get_model_name</span></code> to <code class="docutils literal notranslate"><span class="pre">&quot;ProteomicsRandomForest&quot;</span></code>.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="k">class</span><span class="w"> </span><span class="nc">ProteomicsRandomForest</span><span class="p">(</span><span class="n">RandomForest</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;RandomForest model for drug response prediction using proteomics data.&quot;&quot;&quot;</span>

    <span class="n">cell_line_views</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;proteomics&quot;</span><span class="p">]</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">get_model_name</span><span class="p">(</span><span class="bp">cls</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the model name.</span>

<span class="sd">        :returns: ProteomicsRandomForest</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s2">&quot;ProteomicsRandomForest&quot;</span>
</pre></div>
</div>
<p>2. Next, we need to implement the <code class="docutils literal notranslate"><span class="pre">load_cell_line_features</span></code> method to load the proteomics features.
We already supply proteomics features in the Zenodo as proteomics.csv. Hence, we can already use our
pre-implemented method <code class="docutils literal notranslate"><span class="pre">load_cell_line_features</span></code>.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">load_cell_line_features</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">dataset_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">FeatureDataset</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Loads the cell line features.</span>

<span class="sd">    :param data_path: Path to the gene expression and landmark genes</span>
<span class="sd">    :param dataset_name: Name of the dataset</span>
<span class="sd">    :returns: FeatureDataset containing the cell line proteomics features, filtered through the landmark genes</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">load_and_select_gene_features</span><span class="p">(</span>
        <span class="n">feature_type</span><span class="o">=</span><span class="s2">&quot;proteomics&quot;</span><span class="p">,</span>
        <span class="n">gene_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">data_path</span><span class="o">=</span><span class="n">data_path</span><span class="p">,</span>
        <span class="n">dataset_name</span><span class="o">=</span><span class="n">dataset_name</span><span class="p">,</span>
    <span class="p">)</span>
</pre></div>
</div>
<p>3. We use the same build_model method as the RandomForest class, so we don’t need to implement it.
However, we need to write the hyperparameters needed into the <code class="docutils literal notranslate"><span class="pre">models/baselines/hyperparameters.yaml</span></code> file:</p>
<div class="highlight-YAML notranslate"><div class="highlight"><pre><span></span><span class="nt">ProteomicsRandomForest</span><span class="p">:</span>
<span class="w">  </span><span class="nt">n_estimators</span><span class="p">:</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">100</span>
<span class="w">  </span><span class="nt">max_depth</span><span class="p">:</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">30</span>
<span class="w">  </span><span class="nt">max_samples</span><span class="p">:</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.2</span>
<span class="w">  </span><span class="nt">n_jobs</span><span class="p">:</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">-1</span>
<span class="w">  </span><span class="nt">criterion</span><span class="p">:</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">squared_error</span>
</pre></div>
</div>
<p>We also use the same train and predict method as the RandomForest class, so we don’t need to implement them.</p>
<ol class="arabic simple" start="4">
<li><p>Finally, we need to register the model in the <code class="docutils literal notranslate"><span class="pre">__init__.py</span></code> file in the <code class="docutils literal notranslate"><span class="pre">models/baselines</span></code> directory.</p></li>
</ol>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;MULTI_DRUG_MODEL_FACTORY&quot;</span><span class="p">,</span>
    <span class="s2">&quot;SINGLE_DRUG_MODEL_FACTORY&quot;</span><span class="p">,</span>
    <span class="s2">&quot;MODEL_FACTORY&quot;</span><span class="p">,</span>
    <span class="s2">&quot;NaivePredictor&quot;</span><span class="p">,</span>
    <span class="c1">#[...]</span>
    <span class="s2">&quot;DIPKModel&quot;</span><span class="p">,</span>
    <span class="s2">&quot;ProteomicsRandomForest&quot;</span>
<span class="p">]</span>
<span class="c1">#[...]</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">.baselines.sklearn_models</span><span class="w"> </span><span class="kn">import</span> <span class="p">(</span>
    <span class="n">ElasticNetModel</span><span class="p">,</span> <span class="n">GradientBoosting</span><span class="p">,</span> <span class="n">RandomForest</span><span class="p">,</span> <span class="n">SVMRegressor</span><span class="p">,</span> <span class="n">ProteomicsRandomForest</span>
<span class="p">)</span>

<span class="c1"># SINGLE_DRUG_MODEL_FACTORY is used in the pipeline!</span>
<span class="n">SINGLE_DRUG_MODEL_FACTORY</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">type</span><span class="p">[</span><span class="n">DRPModel</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span>
    <span class="c1">#[...]</span>
<span class="p">}</span>

<span class="c1"># MULTI_DRUG_MODEL_FACTORY is used in the pipeline!</span>
<span class="n">MULTI_DRUG_MODEL_FACTORY</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">type</span><span class="p">[</span><span class="n">DRPModel</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;NaivePredictor&quot;</span><span class="p">:</span> <span class="n">NaivePredictor</span><span class="p">,</span>
    <span class="c1">#[...]</span>
    <span class="s2">&quot;DIPK&quot;</span><span class="p">:</span> <span class="n">DIPKModel</span><span class="p">,</span>
    <span class="s2">&quot;ProteomicsRandomForest&quot;</span><span class="p">:</span> <span class="n">ProteomicsRandomForest</span><span class="p">,</span>
<span class="p">}</span>
</pre></div>
</div>
<ol class="arabic simple" start="5">
<li><p>Add your model to the tests, in this case in <code class="docutils literal notranslate"><span class="pre">tests/individual_models/test_baselines.py</span></code>.</p></li>
</ol>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="nd">@pytest</span><span class="o">.</span><span class="n">mark</span><span class="o">.</span><span class="n">parametrize</span><span class="p">(</span>
    <span class="s2">&quot;model_name&quot;</span><span class="p">,</span>
    <span class="p">[</span>
        <span class="s2">&quot;NaivePredictor&quot;</span><span class="p">,</span>
        <span class="s2">&quot;NaiveDrugMeanPredictor&quot;</span><span class="p">,</span>
        <span class="s2">&quot;NaiveCellLineMeanPredictor&quot;</span><span class="p">,</span>
        <span class="s2">&quot;NaiveMeanEffectsPredictor&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ElasticNet&quot;</span><span class="p">,</span>
        <span class="s2">&quot;RandomForest&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SVR&quot;</span><span class="p">,</span>
        <span class="s2">&quot;MultiOmicsRandomForest&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GradientBoosting&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ProteomicsRandomForest&quot;</span><span class="p">,</span>
    <span class="p">],</span>
<span class="p">)</span>
<span class="nd">@pytest</span><span class="o">.</span><span class="n">mark</span><span class="o">.</span><span class="n">parametrize</span><span class="p">(</span><span class="s2">&quot;test_mode&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;LPO&quot;</span><span class="p">,</span> <span class="s2">&quot;LCO&quot;</span><span class="p">,</span> <span class="s2">&quot;LDO&quot;</span><span class="p">])</span>
<span class="k">def</span><span class="w"> </span><span class="nf">test_baselines</span><span class="p">(</span>
    <span class="n">sample_dataset</span><span class="p">:</span> <span class="n">DrugResponseDataset</span><span class="p">,</span>
    <span class="n">model_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">test_mode</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">cross_study_dataset</span><span class="p">:</span> <span class="n">DrugResponseDataset</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="c1"># [...]</span>
</pre></div>
</div>
<p>6. Now, we add the appropriate documentation.
In our case, the class methods etc. under API are rendered automatically because it is a subclass of Sklearn Models.
If you implement a new model, please orient yourself on the documentation of, e.g., DIPK.</p>
<p>Add your model in <code class="docutils literal notranslate"><span class="pre">usage.rst</span></code> under the section “Available models”.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="usage.html" class="btn btn-neutral float-left" title="How to use DrEvalPy" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="contributing.html" class="btn btn-neutral float-right" title="Contributor Guide" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, DaiSyBio at Technical University of Munich.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>